test-gui/src/package/view-decoded-linear-position-plot/DecodedLinearPositionDrawing.ts

```diff
@@ -0,0 +1,213 @@
+        const uncopyableRightWidth = Math.max(0, targetEnd - currentEnd)
+        // The copy has to happen first, or we'll start overwriting the data we want to keep.
+        if (copyRange[0] < copyRange[1]) {
+            const copyStart = copyRange[0] - currentStart
+            const copyTarget = uncopyableLeftWidth
+            c.drawImage(canvas, copyStart, 0, copyWidth, canvas.height, copyTarget, 0, copyWidth, canvas.height)
+        }
+        if (uncopyableLeftWidth > 0) {
+            painter(targetStart, uncopyableLeftWidth, 0, sampledData, c)
+        }
+        if (uncopyableRightWidth > 0) {
+            painter(targetEnd - uncopyableRightWidth, uncopyableRightWidth, canvas.width - uncopyableRightWidth - 1, sampledData, c)
+        }
+    }
+}
+
+type OffscreenPainter = (startInclusive: number, width: number, pixelXOffset: number, sampleData: DownsampledData, c: CanvasRenderingContext2D) => void
+export const useOffscreenPainter = (styles: string[], height: number, myPositions: number[]) => {
+    const drawTimeIndex = useCallback((data: DownsampledData, firstDataIndex: number, timeIndex: number, xPosition: number, c: CanvasRenderingContext2D) => {
+        const lastIndex = firstDataIndex + data.downsampledTimes[timeIndex]
+        for (let dataIndex = firstDataIndex; dataIndex < lastIndex; dataIndex++) {
+            const position = data.downsampledPositions[dataIndex]
+            const value = data.downsampledValues[dataIndex]
+            c.strokeStyle = styles[value]
+            c.beginPath()
+            const startPixel = myPositions[position]
+            const endPixel = myPositions[position + 1]
+            c.moveTo(xPosition, height - startPixel)
+            c.lineTo(xPosition, height - endPixel)
+            c.stroke()
+        }
+    }, [styles, height, myPositions])
+
+    const clearRect = useCallback((width: number, pixelOffset: number, c: CanvasRenderingContext2D) => {
+        c.fillStyle = styles[0]
+        c.fillRect(pixelOffset, 0, width, height)
+    }, [styles, height])
+
+    const offscreenPainter = useCallback((startInclusive: number, width: number, pixelXOffset: number, sampledData: DownsampledData, c: CanvasRenderingContext2D) => {
+        if (startInclusive + width >= sampledData.downsampledTimes.length) {
+            console.warn(`offscreenPainter called with end (${startInclusive + width} ${startInclusive} ${width}) outside the data range (${sampledData.downsampledTimes.length}).`)
+            return
+        }
+        clearRect(width, pixelXOffset, c)
+        let dataIndex = 0
+        for (let timeIndex = 0; timeIndex < startInclusive; timeIndex++) {
+            dataIndex += sampledData.downsampledTimes[timeIndex]
+        }
+
+        for (let timeIndex = startInclusive; timeIndex < startInclusive + width; timeIndex++) {
+            const xPosition = timeIndex - startInclusive + pixelXOffset
+            drawTimeIndex(sampledData, dataIndex, timeIndex, xPosition, c)
```

Comment:
Decided to keep the drawing operations (in a function) separate from iterating over the data set--I think this winds up being a bit more readable.

I've also tweaked so that we compute the start of the data indices for the time window to draw, then only go from that data index through the end of the drawing window--saves having to do the test at every iteration & having to spin through the rest of the list once we've finished the part we care about drawing.

(I measured it a few times and found negligible performance impact either way.)